### ISSUE TITLE
torch.det behaviour for large batch sizes on cuda

### ISSUE THREAD

**Issue by necludov** [pytorch/pytorch#73622]
### 🐛 Describe the bug

torch.det() behaviour depends on the batch size when executed on GPU.
```
>>> import torch
>>> A = torch.tensor([[1e-40, 1e-40], [0.0, 0.0]])
>>> B = torch.tile(A, [9,1,1])
>>> print(torch.det(B))
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0.])
>>> B = torch.tile(A, [100,1,1])
>>> print(torch.det(B))
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0.])
>>> C = torch.tile(A, [8,1,1])
>>> C = C.cuda()
>>> print(torch.det(C))
tensor([0., 0., 0., 0., 0., 0., 0., 0.], device='cuda:0')
>>> D = torch.tile(A, [9,1,1])
>>> D = D.cuda()
>>> print(torch.det(D))
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:0')
>>> E = torch.tile(A, [100,1,1])
>>> E = E.cuda()
>>> print(torch.det(E))
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,nan,
        nan, nan, nan, nan], device='cuda:0')
```

### Versions

PyTorch version: 1.10.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-163-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: TITAN Xp
Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] hamiltorch==0.4.0.dev1
[pip3] numpy==1.21.2
[pip3] torch==1.10.1
[pip3] torchaudio==0.10.1
[pip3] torchvision==0.11.2
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] hamiltorch                0.4.0.dev1               pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.21.2           py39h20f2e39_0
[conda] numpy-base                1.21.2           py39h79a1101_0
[conda] pytorch                   1.10.1          py3.9_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.10.1               py39_cu102    pytorch
[conda] torchvision               0.11.2               py39_cu102    pytorch

**Comment by dagitses** [comment]
Forgive my daftness, but why do you think the issue is related to the batch size?

**Comment by necludov** [comment]
I'm sorry for the short description. Indeed, it's maybe related not to the batch size but to the total size of the input tensor. However, I've made this assumption since the function works properly for the tensor C of size [8,2,2] but fails on the tensor D of size [9,2,2] and also fails for larger sizes like [100,2,2].

**Comment by Lezcano** [comment]
This is a MAGMA error. For batches of size > 8 we use MAGMA in the `lu_factor` that we use inside of `linalg_det`. MAGMA does not seem to handle matrices with small values as well as cuSolver.

I suggest we benchmark again whether MAGMA is still faster than cusolver for this operation and update the heuristic accordingly

Then, we can see whether MAGMA is still faster than cuSolver at all (it might very well not be). In any case, with that data we should be able to figure out what's the best way forward.

**Comment by Lezcano** [comment]
@necludov In the meantime, you can use `torch.backends.cuda.preferred_linalg_library("cusolver")` to default to the cuSolver backend.
Remember that this is a global flag, and after performing this operation you'll need to reset it to the `"default"` backend.

**Comment by necludov** [comment]
Thank you, @Lezcano, for an amazingly detailed and swift response! My version doesn't seem to have `torch.backends.cuda.preferred_linalg_library()`, but I'll figure that out. Thanks!

**Comment by ngimel** [comment]
It has been recently added, so you'd need nighlty/1.11 rc binaries to get it

**Comment by Lezcano** [comment]
Actually, on a closer look, what's going on here is that cublas does not handle properly subnormal numbers.

Note that the smallest non-subnormal floating point number is `1.175494 × 10-38`. As such `10-40` is approximated to a subnormal number. `lu_factor` (or `getrf` really) in cublas / cusolver checks whether this number is zero. It is not. Then they proceed to compute `0/x` where `x` is the subnormal number. This returns `NaN`.

As annoying as this is, I vote wont-fix. To fix this, we would have to degrade the performance of all the calls to `lu_factor` (hence all the calls not only to `det` but also to `linalg.solve`) just to account for this edge case. Note that the fix would account for calling `nan_to_num` on this function 

One way to fix this issue from the user side is to check for subnormal numbers in the input and zero them out with `torch.where`.

What do you think @ngimel @mruberry ?

cc @xwang233 regarding the cublas / cusolver bug (it happens in both `getrf` and `getrfBatched`)

**Comment by Lezcano** [comment]
Of course, the other way to mitigate this problem is to cast the inputs to `torch.double` before running these functions. This is very much recommended if precision is important (and not so much speed).

**Comment by mruberry** [comment]
Won't fix seems reasonable to me -- @xwang233 what do you think?

**Comment by xwang233** [comment]
Hi @mruberry , I think "Won't fix" seems reasonable to me as well. Thank you.

**Comment by mruberry** [comment]
Thanks for the review, all. Definitely seems like a gotcha, but working in double precision seems like an appropriate workaround for these numeric issues.
